File: solution_zzj/198._House_Robber.cc
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size()==1) return nums[0];
        vector<int> ans = {nums[0], max(nums[0], nums[1])};
        for (int i=2;i<nums.size();i++) {
            ans.push_back(max(ans[i-2]+nums[i], ans[i-1]));
        }
        return ans[nums.size()-1];
    }
};

File: Weijia/198. House Robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        @cache
        #max money for house i
        def r(i):
            if i ==0:
                return nums[0]
            elif i ==1:
                return max(nums[0],nums[1])
            else:
                return max(r(i-1),r(i-2)+nums[i])
        return r(len(nums)-1)
